src/sync/once.rs

```diff
@@ -0,0 +1,147 @@
+/// for one-time initialization for FFI or related functionality. This type can only be constructed
+/// with [`Once::new()`].
+#[derive(Debug)]
+pub struct Once {
+    // We use the address of the `Once` as an identifier, so it can't be zero-sized even though all
+    // its state is stored in ExecutionState storage
+    _dummy: usize,
+}
+
+/// A `Once` cell can either be `Running`, in which case a `Mutex` mediates racing threads trying to
+/// invoke `call_once`, or `Complete` once an initializer has completed, in which case the `Mutex`
+/// is no longer necessary.
+enum OnceInitState {
+    Running(Rc<Mutex<bool>>),
+    Complete(VectorClock),
+}
+
+impl std::fmt::Debug for OnceInitState {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::Running(_) => write!(f, "Running"),
+            Self::Complete(_) => write!(f, "Complete"),
+        }
+    }
+}
+
+impl Once {
+    /// Creates a new `Once` value.
+    pub const fn new() -> Self {
+        Self { _dummy: 0 }
+    }
+
+    /// Performs an initialization routine once and only once. The given closure will be executed
+    /// if this is the first time `call_once` has been called, and otherwise the routine will *not*
+    /// be invoked.
+    ///
+    /// This method will block the calling thread if another initialization routine is currently
+    /// running.
+    ///
+    /// When this function returns, it is guaranteed that some initialization has run and completed
+    /// (it may not be the closure specified).
+    pub fn call_once<F>(&self, f: F)
+    where
+        F: FnOnce(),
+    {
+        let lock = ExecutionState::with(|state| {
+            // Initialize the state of the `Once` cell if we're the first thread to try
+            if self.get_state(state).is_none() {
+                self.init_state(state, OnceInitState::Running(Rc::new(Mutex::new(false))));
+            }
+
+            let init = self.get_state(state).expect("must be initialized by this point");
+            let init_state = init.borrow();
+            trace!(state=?init_state, "call_once on cell {:p}", self);
+            match &*init_state {
+                OnceInitState::Complete(clock) => {
+                    // If already complete, just update the clock from the thread that inited
+                    let clock = clock.clone();
+                    drop(init_state);
+                    state.update_clock(&clock);
+                    None
+                }
+                OnceInitState::Running(lock) => Some(Rc::clone(lock)),
+            }
+        });
+
+        // If there's a lock, then we need to try racing on it to decide who gets to run their
+        // initialization closure.
+        if let Some(lock) = lock {
+            let mut flag = lock.lock().unwrap();
+            if *flag {
+                return;
+            }
+
+            trace!("won the call_once race for cell {:p}", self);
+            f();
```

Comment:
The doc for `Once` says that if `f` panics here, the `Once` instance gets poisoned.  In this version, Shuttle will just report a panic?  Could you add a test for the "f panics" case?

Reply:
I ended up just implementing poisoning -- it's actually really easy because our `Once` uses a `Mutex` under the hood, and `Mutex` already supports poisoning.